### PR TITLE
Fix message history limiter for tool call

### PR DESCRIPTION
**Pull request by GaoxiangLuo**
## Why are these changes needed?

It will return an error
> BadRequestError: Error code: 400 - {'error': {'message': "Invalid parameter: messages with role 'tool' must be a response to a preceeding message with 'tool_calls'.", 'type': 'invalid_request_error', 'param': 'messages.[1].role', 'code': None}}

because OpenAI API has to takes in pairs of `tool_calls` and `tool`.

The following is a minimal example to reproduce the error.

```python
import os
from autogen import ConversableAgent
from typing import Annotated, Literal
from autogen.agentchat.contrib.capabilities import transform_messages, transforms
from autogen.agentchat.contrib.multimodal_conversable_agent import MultimodalConversableAgent

# Define a tool
Operator = Literal["+", "-", "*", "/"]
def calculator(a: int, b: int, operator: Annotated[Operator, "operator"]) -> int:
    if operator == "+":
        return a + b
    elif operator == "-":
        return a - b
    elif operator == "*":
        return a * b
    elif operator == "/":
        return int(a / b)
    else:
        raise ValueError("Invalid operator")

# Initialize agents
assistant = ConversableAgent(
    name="Assistant",
    system_message="You are a helpful AI assistant. "
    "You can help with simple calculations. "
    "Return 'TERMINATE' when the task is done.",
    llm_config={"config_list": [{"model": "gpt-4", "api_key": os.environ["OPENAI_API_KEY"]}]},
)
user_proxy = ConversableAgent(
    name="User",
    llm_config=False,
    is_termination_msg=lambda msg: msg.get("content") is not None and "TERMINATE" in msg["content"],
    human_input_mode="NEVER",
)

# Register the tool
assistant.register_for_llm(name="calculator", description="A simple calculator")(calculator)
user_proxy.register_for_execution(name="calculator")(calculator)

# Add MessageHistoryLimiter
context_handling = transform_messages.TransformMessages(
    transforms=[
        transforms.MessageHistoryLimiter(max_messages=7), # Set it to 7 that will seperate role "tool_calls" and "tool"
    ]
)
context_handling.add_to_agent(assistant)
# Start the chat
chat_result = user_proxy.initiate_chat(assistant, message="What is (44232 + 13312 / (232 - 32)) * 5?", cache=None)
```

## Related issue number

#3121

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by GaoxiangLuo** [comment]
@microsoft-github-policy-service agree

**Comment by marklysze** [comment]
Thanks @GaoxiangLuo for putting this together and for fixing this error.

I've tested the transform code as is and it does work. However, in testing I have identified that by truncating the messages more (e.g. `max_messages=5`) the original `user` message gets removed and then the LLM can no longer understand the task and terminates early. I think this will be a general issue with the use of this history limiter (separate to your tools issue but also affecting it).

I'm proposing a change to the MessageHistoryLimiter that now includes an additional parameter called `keep_first_message` that will always keep the first message and then concatenate the truncated messages. This ensures that the first message, which typically contains core task information, is kept in the conversation.

Let me know what you think.

Change class MessageHistoryLimiter's `__init__` function:
```
    def __init__(self, max_messages: Optional[int] = None, keep_first_message: Optional[bool] = False):
        """
        Args:
            max_messages Optional[int]: Maximum number of messages to keep in the context. Must be greater than 0 if not None.
            keep_first_message Optional[bool]: Whether to keep the original first message in the conversation history.
                Defaults to False. Does not count towards truncation.
        """
        self._validate_max_messages(max_messages)
        self._max_messages = max_messages
        self._keep_first_message = keep_first_message
```
and the `apply_transform` updated (with your code followed by new code):
```
    def apply_transform(self, messages: List[Dict]) -> List[Dict]:
        """Truncates the conversation history to the specified maximum number of messages.

        This method returns a new list containing the most recent messages up to the specified
        maximum number of messages (max_messages). If max_messages is None, it returns the
        original list of messages unmodified.

        Args:
            messages (List[Dict]): The list of messages representing the conversation history.

        Returns:
            List[Dict]: A new list containing the most recent messages up to the specified maximum.
        """

        if self._max_messages is None:
            return messages

        truncated_messages = messages[-self._max_messages :]
        # If the last message is a tool message, include its preceding message that must be a tool_calls message
        if truncated_messages[0].get("role") == "tool":
            start_index = max(-self._max_messages - 1, -len(messages))
            truncated_messages = messages[start_index:]

        # Keep the first message if required
        if self._keep_first_message and messages[0] != truncated_messages[0]:
            truncated_messages = [messages[0]] + truncated_messages

        return truncated_messages
```

Let me know what you think and whether you want to include it in this change, otherwise I'll create a new PR.

**Comment by GaoxiangLuo** [comment]
Thank you @marklysze. LGTM. Please include it in this change.

**Comment by marklysze** [comment]
> Thank you @marklysze. LGTM. Please include it in this change.

Great, thanks @GaoxiangLuo, I've updated the branch to include those changes. It would worth testing and let us know how you go :).

**Review by WaelKarkoub — COMMENTED** [review]
Thank you @GaoxiangLuo @marklysze for these fixes! I'm just worried that the total count of messages, after applying this transform, becomes unintuitive. For example, if I set `max_messages = 5, keep_first_message = True`, intuitively I would think I would have a list of 5 messages with the first being the first created message. But with this fix, I believe I get back 6 messages.

What do you think of using a queue and just anchoring the important messages (first message, tool call etc..) to respect the`max_messages` argument? Everything else looks great and I'll let you guys decide on the final solution!

File: autogen/agentchat/contrib/capabilities/transforms.py [hunk 1]
```diff
@@ -53,13 +53,16 @@ class MessageHistoryLimiter:
     It trims the conversation history by removing older messages, retaining only the most recent messages.
     """
 
-    def __init__(self, max_messages: Optional[int] = None):
+    def __init__(self, max_messages: Optional[int] = None, keep_first_message: Optional[bool] = False):
```
[the remark this answers is not in the document]
Comment (marklysze) [reply autogen/agentchat/contrib/capabilities/transforms.py:56]: Hey @WaelKarkoub! Hope you're well...

I think it's a fair point regarding the first message not being included in the max_messages, I'll update the code to cover that.

P.s. and will remove the `Optional` for keep_first_message, thanks!

[the remark this answers is not in the document]
Comment (marklysze) [reply autogen/agentchat/contrib/capabilities/transforms.py:56]: I'll commit some changes, just a note that the code will not exceed the `max_messages`. If the target `max_messages` means there would be an orphaned 'tools' message, that won't be included, so you'll end up with one less than the `max_messages`.

[the remark this answers is not in the document]
Comment (marklysze) [reply autogen/agentchat/contrib/capabilities/transforms.py]: Okay, I've committed the update. Have also updated the tests to cater for the historic change that moved some functions to transform_utils and added test for the keep_first_message parameter.

File: autogen/agentchat/contrib/capabilities/transforms.py [hunk 2]
```diff
@@ -53,13 +53,16 @@ class MessageHistoryLimiter:
     It trims the conversation history by removing older messages, retaining only the most recent messages.
     """
 
-    def __init__(self, max_messages: Optional[int] = None):
+    def __init__(self, max_messages: Optional[int] = None, keep_first_message: bool = False):
```
Comment (giorgossideris) [inline autogen/agentchat/contrib/capabilities/transforms.py:56]: Hello, as @marklysze pointed out `keep_first_message` seems necessary in order to ensure that the task is understood and the initial message not removed. Currently, the code alone does not provide clear guidance on when to set this parameter to `True`. Therefore, additional documentation on this aspect would be beneficial to clarify its usage.

**Review by WaelKarkoub — APPROVED** [review]
Looks good!

File: autogen/agentchat/contrib/capabilities/transforms.py (hunk 2 above)
Comment (marklysze) [reply autogen/agentchat/contrib/capabilities/transforms.py:56]: I'll mark this as resolved for now. We can add another PR to enhance if needed.

**Review by marklysze — APPROVED** [review]

**Comment by marklysze** [comment]
> Looks good!

Thanks @WaelKarkoub!

**Review by marklysze — APPROVED** [review]

**Comment by marklysze** [comment]
CC @qingyun-wu, @sonichi - I think we're good to go on this one.

**Comment by milkmeat** [comment]
[autogen/agentchat/contrib/capabilities/transforms.py](https://github.com/microsoft/autogen/blob/0.2/autogen/agentchat/contrib/capabilities/transforms.py)

In MessageHistoryLimiter, line 99 of the file.
It should take _keep_first_message into consider. otherwise when keep_first_message==False , the last message will be inserted in a wrong place.

```
        # Loop through messages in reverse
        for i in range(len(messages) - 1, 0, -1):
            if remaining_count > 1:
                truncated_messages.insert(1 if self._keep_first_message else 0, messages[i])
            if remaining_count == 1:
                # If there's only 1 slot left and it's a 'tools' message, ignore it.
                if messages[i].get("role") != "tool":
-                   truncated_messages.insert(1, messages[i])
+                   truncated_messages.insert(1 if self._keep_first_message else 0, messages[i])
```

for example, following code output is in wrong order
```
from autogen.agentchat.contrib.capabilities import transform_messages, transforms


limiter=transforms.MessageHistoryLimiter(max_messages=3, keep_first_message=False)

limiter.apply_transform([{'role':'user', 'name':'agent', 'content':'0'},
                         {'role':'user', 'name':'agent', 'content':'1'},
                         {'role':'user', 'name':'agent', 'content':'2'},
                         {'role':'user', 'name':'agent', 'content':'3'},
                         {'role':'user', 'name':'agent', 'content':'4'},
                        ])
```
```
[{'role': 'user', 'name': 'agent', 'content': '3'},
 {'role': 'user', 'name': 'agent', 'content': '2'},
 {'role': 'user', 'name': 'agent', 'content': '4'}]
```

**Comment by ekzhu** [comment]
@milkmeat would you like to submit a fix for this as a PR?

**Comment by milkmeat** [comment]
I made a pull request for the bug I mentioned above.
https://github.com/microsoft/autogen/pull/4653

I am new to this, if anything needs to change please let me know.
* I make changes the "0.2" branch, I hope it is correct
* I can't find the place to request a review, @ekzhu can you take a look? 
* I just figured that I am too slow. you alread fixed in in #4644, just feel free to close mine. :)

**Comment by ekzhu** [comment]
@milkmeat hey no worries thanks for your PR #4653. Let's continue our discussion there. I created the issue but haven't done anything.

**Pull request merged**